src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs

```diff
@@ -38,22 +38,266 @@ private unsafe int JsCompareString(ReadOnlySpan<char> string1, ReadOnlySpan<char
             return cmpResult;
         }
 
-        private static bool CompareOptionsNotSupported(CompareOptions options) =>
+        private void JsInit(string interopCultureName)
+        {
+            _isAsciiEqualityOrdinal = GetIsAsciiEqualityOrdinal(interopCultureName);
+        }
+
+        private unsafe int JsIndexOfCore(ReadOnlySpan<char> source, ReadOnlySpan<char> target, CompareOptions options, int* matchLengthPtr, bool fromBeginning)
+        {
+            Debug.Assert(!GlobalizationMode.Invariant);
+            Debug.Assert(!GlobalizationMode.UseNls);
+            Debug.Assert(GlobalizationMode.Hybrid);
+            Debug.Assert(target.Length != 0);
+
+            if (IndexingOptionsNotSupported(options) || ComparisonOptionsNotSupported(options))
+                throw new PlatformNotSupportedException(GetPNSE(options));
+
+            string cultureName = m_name;
+
+            if (ComparisonOptionsNotSupportedForCulture(options, cultureName))
+                throw new PlatformNotSupportedException(GetPNSEForCulture(options, cultureName));
+
+            string exceptionMessage;
+            int idx;
+            if (_isAsciiEqualityOrdinal && CanUseAsciiOrdinalForOptions(options))
+            {
+                idx = (options & CompareOptions.IgnoreCase) != 0 ?
+                    IndexOfOrdinalIgnoreCaseHelperJS(out exceptionMessage, source, target, options, matchLengthPtr, fromBeginning) :
+                    IndexOfOrdinalHelperJS(out exceptionMessage, source, target, options, matchLengthPtr, fromBeginning);
+            }
+            else
+            {
+                fixed (char* pSource = &MemoryMarshal.GetReference(source))
+                fixed (char* pTarget = &MemoryMarshal.GetReference(target))
+                {
+                    idx = Interop.JsGlobalization.IndexOf(out exceptionMessage, m_name, pTarget, target.Length, pSource, source.Length, options, matchLengthPtr, fromBeginning);
+                }
+            }
+
+            if (!string.IsNullOrEmpty(exceptionMessage))
+                throw new Exception(exceptionMessage);
+
+            return idx;
+        }
+
+        /// <summary>
+        /// Duplicate of IndexOfOrdinalHelperJS that also handles ignore case. Can't converge both methods
+        /// as the JIT wouldn't be able to optimize the ignoreCase path away.
+        /// </summary>
+        /// <returns></returns>
+        // ToDo: clean up with IndexOfOrdinalHelper from .Icu
+        private unsafe int IndexOfOrdinalIgnoreCaseHelperJS(out string  exceptionMessage, ReadOnlySpan<char> source, ReadOnlySpan<char> target, CompareOptions options, int* matchLengthPtr, bool fromBeginning)
+        {
```

Comment:
Wouldn't it better to move this function https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs#L95 to CompareInfo.cs and call from ICU and WebAssembly? From first look implementations look identical.

Reply:
It differs with one goto: `InteropCall`. I tried to merge these two implementations: icu + wasm by passing the delegate to the `InteropCall` but unfortunately delegates do not support unsafe parameters (pointers), so the cleanup failed. Generally all `CompareInfo` should get refactored because it's a huge class and I was thinking about doing it in follow-up, when collations on HybridGlobalization would be already working.

Reply:
Also I have some concerns about refactoring these functions. Originally, ICU has 2 very similar functions that differ only with lines
```
                        if (char.IsAsciiLetterLower(valueChar))
                            valueChar = (char)(valueChar - 0x20);
                        if (char.IsAsciiLetterLower(targetChar))
                            targetChar = (char)(targetChar - 0x20);
```
but they have this comment at the top:
```
        /// Duplicate of IndexOfOrdinalHelper that also handles ignore case. Can't converge both methods
        /// as the JIT wouldn't be able to optimize the ignoreCase path away.
```
An attempt to merge Wasm + Icu functions is similar to the situation described. @ViktorHofer, you are the author of the comment, could you help us? Can you think about an effective way of refactoring e.g. `IndexOfOrdinalIgnoreCaseHelper` and `IndexOfOrdinalIgnoreCaseHelperJS` into one function, when they differ only with the contents of `goto InteropCall` instruction?